src/main/java/dk/cngroup/trainings/spring/springassignment/controller/rest/AnimalController.java

```diff
@@ -80,39 +84,45 @@ public AnimalController(AnimalService animalService) {
 	@RequestMapping(path = "/{id}/careTakers", method = RequestMethod.POST)
 	public ResponseEntity<CareTaker> addNewCareTakerToExistingAnimal(@PathVariable("id") long id,
 																	 @RequestBody @Valid CareTaker careTaker) {
-		if (animalService.getAnimalById(id).isPresent()) {
-			Optional<CareTaker> addedCareTaker = null;
-			try {
-				addedCareTaker = animalService.addNewCareTakerToExistingAnimal(id, careTaker);
-				return new ResponseEntity<>(addedCareTaker.get(), HttpStatus.OK);
-			} catch (InvalidCareTakerException e) {
-				return ResponseEntity.badRequest().build();
-			}
-		} else {
+		Optional<CareTaker> addedCareTaker;
+		try {
+			addedCareTaker = animalService.addNewCareTakerToExistingAnimal(id, careTaker);
+			return new ResponseEntity<>(addedCareTaker.get(), HttpStatus.OK);
+		} catch (InvalidCareTakerException e) {
+			return ResponseEntity.badRequest().build();
+		} catch (AnimalNotFoundException e) {
 			return ResponseEntity.notFound().build();
 		}
 	}
 
 	@RequestMapping(value = "/{id}/careTakers", method = RequestMethod.GET)
 	public ResponseEntity<List<CareTaker>> getCareTakersByAnimalId(@PathVariable("id") long id) {
-		Optional<Animal> animalById = animalService.getAnimalById(id);
-		if (animalById.isPresent()) {
+		Optional<Animal> animalById;
+		try {
+			animalById = animalService.getAnimalById(id);
 			return new ResponseEntity<>(animalById.get().getCareTakers()
 					, HttpStatus.OK);
-		} else {
+		} catch (AnimalNotFoundException e) {
 			return ResponseEntity.notFound().build();
 		}
 	}
 
 	@RequestMapping(path = "/{animalId}/careTakers/{careTakerId}", method = RequestMethod.PUT)
 	public ResponseEntity<String> addExistingCareTakerToExistingAnimal(@PathVariable("animalId") long animalId,
```

Comment:
In REST, you usually return what you have added with a request... so in here, I think you could either return the CareTaker or the Animal. Returning String is a bit unnecessary.

Reply:
I used a String because that case has many results.
In a successful case, it makes sens to return the CareTaker, since we are adding it to an animal.
But in a fail, we have 3 cases: careTaker doesn't exist, animal doesn't exist or the both of them exist and already linked. So returning a plain NotFound response Entity won't give enough description to the client.
But I guess I should follow the norm when it comes to REST web services.

Reply:
I understand what you want to do, well `ResponseEntity` is reaching its limits here. What you could do is to use a Controller Advice, which would handle the errors: https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

So in the Controller, you would return only `Animal` (for example), no response entity, and you would not catch any exceptions either (just add `throws` to the method signature). And you would implement a Controller Advice, which would catch these exceptions and handle them properly. You can try it if you want, but it is fine like this for me at the moment.